typed-protocols/src/Network/TypedProtocol/PingPong/Codec.hs

```diff
@@ -23,28 +23,29 @@ codecPingPong =
     encode (ServerAgency TokBusy) MsgPong = "pong\n"
 
     decode :: PeerHasAgency pr st
-           -> m (DecodeStep String String m (SomeMessage st))
+           -> m (DecodeStep String CodecFailure m (SomeMessage st))
     decode stok =
       decodeTerminatedFrame '\n' $ \str trailing ->
         case (stok, str) of
           (ServerAgency TokBusy, "pong") -> DecodeDone (SomeMessage MsgPong) trailing
           (ClientAgency TokIdle, "ping") -> DecodeDone (SomeMessage MsgPing) trailing
           (ClientAgency TokIdle, "done") -> DecodeDone (SomeMessage MsgDone) trailing
-          _                              -> DecodeFail ("unexpected message: " ++ str)
+          _                              -> DecodeFail failure
+            where failure = CodecFailure ("unexpected message: " ++ str)
```

Comment:
@dcoutts we know the agency and the token at this point, it might be useful to include it in the failure message.